### PR TITLE
Add support for textDocument/foldingRange

### PR DESCRIPTION
**Pull request by rgrinberg**
The functionality is provided using merlin's outline call.

## TODO

- [x] Add tests
- [x] Fold module types
- [x] Fold modules
- [x] Supporting folding range capabilities
- [x] Sort folds in location order
- [x] Fold classes

### CODE REVIEW

File: lsp/src/protocol.ml [hunk 1]
```diff
@@ -7270,3 +7446,285 @@ module DebugTextDocumentGet = struct
...
+        | _ -> yojson_error "invalid t_of_yojson" node )
+      | _ -> yojson_error "invalid t_of_yojson" node
+  end
+
+  type t =
+    { startLine : zero_based_int
+    ; startCharacter : zero_based_int option
+    ; endLine : zero_based_int
+    ; endCharacter : zero_based_int option
+    ; kind : Kind.t
```
[the remark this answers is not in the document]
Comment (rgrinberg) [reply lsp/src/protocol.ml]: Makes sense. Does it also make sense to allow raw string values here? The spec isn't quite clear to me.

File: ocaml-lsp-server/src/ocaml_lsp_server.ml [hunk 2]
```diff
@@ -529,6 +530,36 @@ let on_request :
...
+        ; endCharacter = Some range.end_.character
+        ; kind = Region
+        }
+      in
+      let rec loop acc (items : Query_protocol.item list) =
+        match items with
+        | [] -> acc
+        | item :: items -> (
+          let items = item.children @ items in
+          match item.outline_kind with
```
[the remark this answers is not in the document]
Comment (rgrinberg) [reply ocaml-lsp-server/src/ocaml_lsp_server.ml]: Hmm, perhaps we should just be consistent and just fold anything over 2 lines?

[the remark this answers is not in the document]
Comment (andreypopp) [reply ocaml-lsp-server/src/ocaml_lsp_server.ml]: Maybe, but 2 lines is too small I think. I personally not a heavy user of folding though.

File: lsp/src/protocol.ml (hunk 1 above)
[the remark this answers is not in the document]
Comment (andreypopp) [reply lsp/src/protocol.ml]: I think there's a typo in spec, allowing `Kind.t` is the right thing.

File: ocaml-lsp-server/src/ocaml_lsp_server.ml (hunk 2 above)
[the remark this answers is not in the document]
Comment (rgrinberg) [reply ocaml-lsp-server/src/ocaml_lsp_server.ml]: Well I don't think that anything will be folded by default anyway. It only gives users the option to fold. In any case, we can do something like >=4 lines.

**Comment by rgrinberg** [comment]
Tested it on VSC and it works nicely. In the end, I elected to allow the user to fold anything that is foldable. The user is never forced to fold anything, so why not give them the opportunity whenever it's possible

**Pull request merged**